File: Manh/bai-tap-buoi-so-4-de-quy.hs
{-# LANGUAGE OverloadedStrings #-}
module Example where

import Language.Marlowe.Extended.V1
import Data.String

main :: IO ()
main = printJSON example


{- Define a contract, Close is the simplest contract which just ends the contract straight away
-}

-- Các bên tham gia vào hợp đồng
chuDoanhNghiep = Role "ChuDoanhNghiep"

--Các tham số dùng trong hợp đồng
thoiHanChuyenLuong = TimeParam "ThoiHanChuyenLuong"
soLuongNhanVien = 50

danhSachNhanVien :: [String]
danhSachNhanVien = ["NhanVien" ++ show i | i <- [1..soLuongNhanVien]]

luongToanBoNhanVien = ConstantParam "LuongToanBoNhanVien"
luong1NhanVien = DivValue luongToanBoNhanVien (Constant soLuongNhanVien)

-- hàm này sẽ thực hiện lệnh Pay một cách đệ quy để trả lương cho toàn bộ nhân viên từ tài khoản của chủ doanh nghiệp
traLuong :: [String] -> Contract
traLuong [] = Close
traLuong (nhanVien:cacNhanVienConLai) = Pay chuDoanhNghiep (Party (fromString nhanVien)) ada (UseValue "Luong1NhanVien") (traLuong cacNhanVienConLai)


chuyenTienLuongVaoHopDong :: Contract -> Contract
chuyenTienLuongVaoHopDong buocTiepTheo = 
    When
        [Case
            (Deposit chuDoanhNghiep chuDoanhNghiep ada luongToanBoNhanVien)
            (Let "Luong1NhanVien" luong1NhanVien buocTiepTheo)]
        (TimeParam "ThoiHanChuyenLuong")
        Close 


example :: Contract
example = chuyenTienLuongVaoHopDong (traLuong danhSachNhanVien)

File: Manh/bai-tap-buoi-so-4-de-quy.ts
import {
    Address, Role, Account, Party, ada, lovelace, AvailableMoney, Constant, ConstantParam,
    NegValue, AddValue, SubValue, MulValue, DivValue, ChoiceValue, TimeIntervalStart,
    TimeIntervalEnd, UseValue, Cond, AndObs, OrObs, NotObs, ChoseSomething,
    ValueGE, ValueGT, ValueLT, ValueLE, ValueEQ, TrueObs, FalseObs, Deposit,
    Choice, Notify, Close, Pay, If, When, Let, Assert, SomeNumber, AccountId,
    ChoiceId, Token, ValueId, Value, EValue, Observation, Bound, Action, Payee,
    Case, Timeout, ETimeout, TimeParam, Contract
} from 'marlowe-js';

(function (): Contract {

    // Các bên tham gia vào hợp đồng
    const chuDoanhNghiep : Party = Role("ChuDoanhNghiep");

    // Các tham số dùng trong hợp đồng
    const soLuongNhanVien = 50;
    const thoiHanChuyenLuong : Timeout = TimeParam("ThoiHanChuyenLuong");
    const luongToanBoNhanVien : Value = ConstantParam("LuongToanBoNhanVien");
    const luong1NhanVien : Value = DivValue(luongToanBoNhanVien, Constant(soLuongNhanVien));

    function traLuong (sttNhanVien: number) : Contract {
        if (sttNhanVien === soLuongNhanVien) {
            // Nếu đã đến lượt nhân viên cuối cùng, thì sau khi trả lương xong hợp đồng sẽ Close
            return Pay(chuDoanhNghiep, Party(Role(`NhanVien${sttNhanVien}`)), lovelace, UseValue("Luong1NhanVien"), Close);    
        } else {
            return Pay(chuDoanhNghiep, Party(Role(`NhanVien${sttNhanVien}`)), lovelace, UseValue("Luong1NhanVien"), traLuong(sttNhanVien+1));
        }
    }

    function chuyenTienLuongVaoHopDong (buocTiepTheo: Contract) {
        return When([
            Case(Deposit(chuDoanhNghiep, chuDoanhNghiep, lovelace, luongToanBoNhanVien), 
                Let("Luong1NhanVien", luong1NhanVien, buocTiepTheo)
            )
        ], thoiHanChuyenLuong, Close);
    }

    return chuyenTienLuongVaoHopDong (traLuong (1));
})
